include/deal.II/base/function_bounds.h

```diff
@@ -0,0 +1,112 @@
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_function_bounds_h
+#define dealii_function_bounds_h
+
+
+#include <deal.II/base/config.h>
+
+#include <deal.II/base/bounding_box.h>
+#include <deal.II/base/function.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+/**
+ * Struct storing lower and upper bounds, $[L, U]$, on the value of a
+ * function, $f(x)$, such that
+ *
+ * $L \leq f(x) \leq U$.
+ */
+struct FunctionBounds
+{
+public:
+  /**
+   * Initialize the bounds to $[-\infty, \infty]$,
+   * corresponding to no information about a functions range.
+   */
+  FunctionBounds();
+
+  /**
+   * The lower bound, $L$, on the function value:
+   *
+   * $L \leq f(x)$.
+   */
+  double min;
+
+  /**
+   * The upper bound, $U$, on the function value:
+   *
+   * $f(x) \leq U$.
+   */
+  double max;
+};
+
+
+/**
+ * Estimate bounds on the value and each gradient component of a Function, $f$,
+ * over a BoundingBox, by approximating it as a 2nd order Taylor polynomial
+ * starting from the box center.
+ *
+ * The function value, gradient and Hessian are computed at the box center.
+ * The upper and lower bounds on the functions value are then estimated as
+ *
+ * $f(x) \in [f(x_c) - F, f(x_c) + F]$,
+ * where
+ * $F = \sum_i |\partial_i f(x_c)| h_i
+ *    + 1/2 \sum_i \sum_j |\partial_i \partial_j f(x_c)| h_i h_j$.
+ *
+ * Here, $h_i$ is half the side length of the box, which is the distance we
+ * extrapolate. The bounds on the gradient components are estimated similarly as
+ *
+ * $\partial_i f \in [\partial_i f(x_c) - G_i, \partial_i f(x_c) + G_i]$,
+ * where
+ * $G_i = \sum_j |\partial_i \partial_j f(x_c)| h_j$.
+ *
+ * If the function has more than 1 component the @p component parameter can
+ * be used to specify which function component the bounds should be computed
+ * for.
+ */
+template <int dim>
+void
+taylor_estimate_function_bounds(
+  const Function<dim> &            function,
+  const BoundingBox<dim> &         box,
+  FunctionBounds &                 value_bounds,
+  std::array<FunctionBounds, dim> &gradient_bounds,
+  const unsigned int               component = 0);
+
+
+/**
+ * Return a lower bound, $L_a$, on the absolute value of a function, $f(x)$:
+ *
+ * $L_a \leq |f(x)|$,
+ *
+ * by estimating it from the incoming lower and upper bounds:
+ * $L \leq f(x) \leq U$.
+ *
+ * By rewriting the above as
+ * $F - C \leq f(x) \leq F + C \LeftRightArrow |f(x) - F| \leq C$,
+ * with $F = (U + L)/2$, $C = (U - L)/2$, and using the inverse triangle
+ * inequality, $|F| - |f(x)| \leq |f(x) - F|$, we get that $L_a = |F| - C$.
+ * If this is negative, $L_a$ is set to 0, since the absolute value can't be
+ * negative.
+ */
+double
+lower_bound_on_abs(const FunctionBounds &function_bounds);
```

Comment:
On second thought about this function, I think it is too specialized. I can see the point of the function above, but would suggest that you keep this one where you actually need it. I think that would also address the issues with the documentation: The function needs substantially more space for documentation that for the implementation because one has to explain the context, and if you just inlined the computations this function does into the place where you currently use it, the context is easier to understand for a reader.